SI-8990 Fix lazy evaluation of StreamWithFilter#foreach

ms-tg opened the pull request:
Per discussion with @Ichoran, this fix requires changing the return type of `Stream#withFilter` to _not_ be an inner class of `Stream` in order to avoid holding the reference to the head of the original structure.

Therefore, this PR is rebased on 2.12.x, and is squashed to one commit per CONTRIBUTING.md.

ms-tg committed:
SI-8990 Allow GC during lazy evaluation of Stream#withFilter

  - The fact that StreamWithFilter was an inner class prevented
    collection of the head during processing after #withFilter,
    due to the reference to the outer class instance held in self.
  - By implementing StreamWithFilter outside of the Stream class,
    we gain control of the reference to the original Stream.
  - We clarify explicit "filter after first use" semantics for the
    reference to the original stream, which allows it to be GC'd
    during the processing of foreach, map, and flatMap.
  - Code is more DRY by implementing in terms of Stream#filter,
    which is already correctly lazy, and then Stream's #map,
    #flatMap, and #foreach, which already correctly allow GC.
  - Unfortunately, the type returned by Stream#withFilter *must*
    change, as it had previously inherited from the inner class
    TraversableLike#WithFilter, which is where the problematic
    reference to the outer class instance was held. Therefore,
    this change is targetted to 2.12.x. There doesn't appear to
    be any way to fix this without changing the type, sadly.
  - Special thanks to @paulp who suggested the likely cause of the
    issue at nescala 2015, and got me setup to build and run tests.
    Also thanks to @Ichoran and @retronym, who suggested that
    filter-after-first-use would preserve the reusable nature
    of the return value from #withFilter, rather than a single-shot
    null-after-first-use solution.

Review by @Ichoran and @retronym
Fixes #8990

ms-tg commented:
Why are the tests above failing to launch? And appearing to base against 2.11.x?

ms-tg commented:
/rebuild

ms-tg commented:
Also, does the new (non-inner) class `StreamWithFilter` need to be declared private? If so, private to what package?

ms-tg commented:
/rebuild

adriaanm commented:
Sorry about the brokenness. We're in the process of moving over to new CI. We first implemented it in 2.11.x, and the merge of 2.11.x into 2.12.x (#4272) will need to be merged before 2.12.x CI can succeed. Will take care of that on Monday.

ms-tg commented:
@adriaanm ah ok, thanks for explaining.

retronym commented:
#4272 has now been merged. Can you please rebase on the latest 2.12.x branch now? Sorry for the hassle.

ms-tg commented:
Sure @retronym , I'll rebase and repush now

ms-tg commented:
@retronym: any idea what this failure means? https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-ide/46/console

ms-tg commented:
@Ichoran @retronym: any help understanding these test failures would be much appreciated

ms-tg commented:
/rebuild

adriaanm commented:
@ms-tg, the ide failure is unrelated to your PR -- sorry! I think it's caused by a broken https://github.com/scala-ide/uber-build/blob/master/zinc/sbt-on-2.12.x 

I've submitted to skip IDE validation on 2.12.x for now https://github.com/scala/scala/pull/4287.
We'll ignore this failure on your PR.

Still need to figure out what went wrong with the JUnit test on Stream:

```
Testsuite: scala.collection.immutable.StreamTest
Tests run: 3, Failures: 1, Errors: 0, Time elapsed: 11.949 sec

Testcase: foreach_allows_GC took 0.396 sec
Testcase: withFilter_foreach_allowsGC took 11.515 sec
    FAILED
assertion failed
junit.framework.AssertionFailedError: assertion failed
    at scala.Predef$.assert(Predef.scala:151)
    at scala.collection.immutable.StreamTest.withFilter_foreach_allowsGC(StreamTest.scala:39)

Testcase: t6727_and_t6440 took 0.018 sec
```

From https://scala-ci.typesafe.com/job/scala-2.11.x-validate-test/ws/build/junit/TEST-scala.collection.immutable.StreamTest.txt (will be archived again from now on: https://github.com/scala/scala-jenkins-infra/pull/22)

ms-tg commented:
Thank you @adriaanm ! That test is passing locally and passed on previous CI run. I will run repeatedly locally to verify hypothesis that it is simply an intermittent failure to get the GC to trigger before the test is over, rather than a real failure. I can increase the number of attempts and repetitions to stamp this out, the same as with other Scala library tests of GC on which these were modelled.

adriaanm commented:
Thanks. Spurious failures are a real buzz kill, so we should try to eliminate any uncertainty in these tests.

ms-tg commented:
@adriaanm: I am unable to reproduce such a failure locally. (rest of comment deleted, as it was incorrect -- the failed test was of the new behavior, which I had misread originally)

ms-tg commented:
/rebuild

adriaanm commented:
Maybe it's due to the new CI environment, which now runs on EC2 (c4.2xlarge, so not too shabby).

adriaanm commented:
It seems to fail pretty consistently, though... :-/

ms-tg commented:
@adriaanm: can you help me identify any other aspects of compiler or library code that could be different in this environment than locally?

adriaanm commented:
Try running with these JAVA_OPTS maybe? https://github.com/scala/scala-jenkins-infra/blob/master/attributes/master.rb#L28

adriaanm commented:
We run on the latest Oracle JDK 6

ms-tg commented:
@adriaanm: good point, I'll look at that next. First I'll try running the exact line that failed, `ant -Dstarr.version=2.12.0-7714224-SNAPSHOT -Dscalac.args.optimise=-optimise -Dlocker.skip=1 -Dstarr.use.released=1 -Dextra.repo.url=http://private-repo.typesafe.com/typesafe/scala-pr-validation-snapshots/ test.core docs.done`

ms-tg commented:
@adriaanm I cannot reproduce with the `JAVA_OPTS` and the exact ant command. I am using Java 7 on OS X Yosemite, however.

adriaanm commented:
I'm retrying the failed build under openjdk 7: https://scala-ci.typesafe.com/job/scala-2.11.x-validate-test/77/console (See: https://scala-ci.typesafe.com/job/scala-2.11.x-validate-test/77/parameters/)

adriaanm commented:
/synch

ms-tg commented:
@adriaanm wow, I see that the test `StreamTest#withFilter_foreach_allowsGC` failed on openjdk 7. Can you please help me to understand how to reproduce this locally? It passes every time locally. Are we sure that it is using the Stream class modifications from this PR?

ms-tg commented:
@adriaanm: also, can we run against 163490fe3b28de2413daccb51a17cdce2c9c39f9 instead in order to give it up to 30s to begin GC collection of the head? It will finish much faster if possible.

ms-tg commented:
@adriaanm Success (in failure)! By reproducing what the CI machine did exactly, I produced the same failure. Commands:

```
git checkout 163490fe3b28de2413daccb51a17cdce2c9c39f9
git reset --hard
git clean -fdx
export scalaVersion=2.12.0-7714224-SNAPSHOT
source scripts/jobs/validate/test
```

Can you please clarify the following obvious question: are these commands expected to actually build the the library code and test it? Because the behavior with this code is different from, for example, running `and test.junit`, in which case my test passes?

adriaanm commented:
The test job compiles the library/reflect/compiler with the compiler built by the publish-core job (i.e., bootstraps the compiler). So, yes, the CI job should recompile the library and use that to run the junit tests. The `ant test.junit` runs the test suite directly -- I'm going to defer to @gkossakowski on what this test target is meant to do in terms of compilation, but I assume it tests the library in build/quick/classes/library. Have you tried a `git clean -fxd && ant test.core`?

ms-tg commented:
@adriaanm: I can try to find time to understand the difference, but this is getting beyond me. I just edited the test file to print out what class it was getting at the .withFiter call and re-ran, however this change of code somehow caused it to pass on all subsequent runs, and until I did the git reset and git clean commands again.

Any ideas?

adriaanm commented:
It sounds like the `test.junit` target in the build may not be picking up changes to your source files. The last idea I can come up with is to delete `test-compile.complete` (in `build/junit`, I think) before running test.junit.

ms-tg commented:
@adriaanm: I think I found the issue. Out of sheer frustration, I tried running tests using Scala 2.11.5, which is what I developed this fix against:

```
export scalaVersion=2.11.5
git checkout 163490fe3b28de2413daccb51a17cdce2c9c39f9
git reset --hard
git clean -fdx
source scripts/jobs/validate/test
```

I see that my test now passes!

Since not much in this area of code changed, I have a strong hunch that it is related to https://github.com/scala/scala/pull/3925, https://github.com/scala/scala/pull/3949, and related changes, which implement `#filter` and `#filterNot` through `#filterImpl`. It looks as though the extra stack frame could be what is preventing the head from being GC'd here.

I will investigate further, but was wondering if anyone who worked on those commits could shed any further light? @Ichoran @lrytz @paulp @ruippeixotog @gkossakowski

gkossakowski commented:
Yes, the `ant test.junit` is supposed to test library built into `build/quick/classes/library`. Do you still have problems understanding JUnit failures or it's different issue that is being discussed in last comments?

adriaanm commented:
@ms-tg, Glad you figured it out! Excellent! 
@gkossakowski, I think the issue is now back to the code instead of the build

ms-tg commented:
Excellent. I will make both suggested changes as soon as I get home. If this looks acceptable then I will squash again :)

Ichoran commented:
I think that in this case the test that ended up in https://github.com/scala/scala/pull/4133 (which I just merged into 2.11.x) is probably _not_ what you want.  But if you go through the history you'll see a test with a more strenuous effort to invoke and give time and reason for GC. Perhaps that should be resurrected for cases like this (where it might be difficult to define exactly who is around from which to reach).

retronym commented on src/library/scala/collection/immutable/Stream.scala:
```diff
@@ -1293,6 +1244,23 @@ object Stream extends SeqFactory[Stream] {
+    * Because this is not an inner class of `Stream`, it is now possible to offer explicit
+    * 'null after first use' semantics for the reference to the head of the original
+    * `Stream`, allowing GC to collect while the tail is still processing (see SI-8990).
+    */
```
I don't think we can make filtered Streams single-shot only. `Stream` is immutable, so I would expect the following to offer consistent results and not NPE.

``` scala
% scala-hash v2.11.5
scala> val s = Stream(1, 2, 3, 4)
s: scala.collection.immutable.Stream[Int] = Stream(1, ?)

scala> val filtered = s withFilter (_ % 2 == 0)
filtered: s.StreamWithFilter = scala.collection.immutable.Stream$StreamWithFilter@1002d1c8

scala> filtered.map(identity).force
res7: scala.collection.immutable.Stream[Int] = Stream(2, 4)

scala> filtered.map(identity).force
res8: scala.collection.immutable.Stream[Int] = Stream(2, 4)
```

When the one-shot semantics are needed, `Stream#iterator` is available.

ms-tg replied:
That's an interesting point. So is it not the case that withFilter exists to support for expressions? In that context, they are indeed single shot. And in fact, it is supporting GC in for-expression use cases that prompted this issue.

Of course, if they are not single-shot, then I don't believe they can allow GC -- as long as the filtered collection exists, it must hold a reference to the head of the original one. Do you agree?

retronym replied:
`withFilter` was introduced to reduce intermediate collection creation in for expressions, but the method is available for anyone to call in regular code too.

![image](https://cloud.githubusercontent.com/assets/65551/6054253/a0ebb5f8-ad36-11e4-90a1-d06d8bcb184a.png)

I think you are correct in your assessment that we can't have an immutable StreamWithFilter _and_ avoid retaining the head reference.

Ichoran replied:
You can't avoid the head reference, but in obeying the _spirit_ rather than the letter of the API, you _can_ discard all of the leading and filtered-out elements.  For non-memoizing eager collections, restricting the domain is the thing to do to save memory: you've already got it all, just pick the ones you want.  But for memoizing lazy collections, you want _exactly_ the opposite.  The only difference is that you'd like to throw away the head lazily also, which regular `filter` doesn't support.  So `withFilter`'s job should be to take the original collection, call `filter` on it at most one time, and store that result instead.  That way the intent of `withFilter` is filled: do as little work as possible.

retronym replied:
I had to read that a few times until I understood it. Worth it, though: Clever!

Just to confirm my understanding, do you mean?

``` scala
class StreamWithFilter[A](private var s: Stream[A], pred: A => Boolean) {
   lazy private val filtered = try { s filter pred } finally { s = null }
  def map[B, That](f: A => B)(implicit bf: CanBuildFrom[Stream[A], B, That]): That = filtered.map(f)
  // ...
}
```

Ichoran replied:
@retronym Yup, that's it!  Not sure whether it should be try/finally or a temporary assignment to a val, though.  Depends on whether you think a second attempt at filtering could succeed.  Could argue it either way.

ms-tg replied:
@retronym @Ichoran Ok, I definitely can get behind this propososal!

I will change this PR to accomplish that and to test it. However, after I do so, can I ask for some help in getting the collectability to work when Scala 2.12.x optimizing compiler is run? ;)

Ichoran commented:
I am happy to try to help with intermittent collection failures, but it will have to wait until the 10th; I am too busy until then.

retronym commented:
Looks like I gave you slightly incorrect advice in my email about bytecode diffing.

To compile the JUnit tests with -optimize:

```
% ant -Dscalac.args.optimise=-optimize test.junit
```

Here's the bytecode diff (and instructions to produce it):

https://gist.github.com/retronym/88dc746a4e09232a9917#file-streamtest-decompiled-diff-L913

retronym commented:
Here is a minimized version of what we're looking at:

``` scala
object Test {
  def main(args: Array[String]): Unit = {
    test()
  }
  def test() {
    new Stream().withFilter.next.next.next.next.next.next.next.next
  }
  class Stream() {
    println("new Stream()")
    val data = Array.ofDim[Byte](64 * 1024 * 1024)
    lazy val next = new Stream
    @inline
    final def withFilter = new StreamWithFilter(this)
  }

  final class StreamWithFilter(var s: Stream) {
    lazy val filtered = { val result = s.next; s = null; result}  
    def next: Stream = filtered.next
  }
}
```

```
% qscalac sandbox/test.scala && echo ':javap -c Test$#test' | qscala && qscala -nc Test
Welcome to Scala version 2.12.0-20150204-161502-f894c55ef3 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_25).
Type in expressions to have them evaluated.
Type :help for more information.

scala> :javap -c Test$#test
  public void test();
    Code:
       0: new           #22                 // class Test$Stream
       3: dup
       4: invokespecial #23                 // Method Test$Stream."<init>":()V
       7: invokevirtual #27                 // Method Test$Stream.withFilter:()LTest$StreamWithFilter;
      10: invokevirtual #33                 // Method Test$StreamWithFilter.next:()LTest$Stream;
      13: invokevirtual #34                 // Method Test$Stream.next:()LTest$Stream;
      16: invokevirtual #34                 // Method Test$Stream.next:()LTest$Stream;
      19: invokevirtual #34                 // Method Test$Stream.next:()LTest$Stream;
      22: invokevirtual #34                 // Method Test$Stream.next:()LTest$Stream;
      25: invokevirtual #34                 // Method Test$Stream.next:()LTest$Stream;
      28: invokevirtual #34                 // Method Test$Stream.next:()LTest$Stream;
      31: invokevirtual #34                 // Method Test$Stream.next:()LTest$Stream;
      34: pop
      35: return
}

scala> :quit
new Stream()
new Stream()
new Stream()
new Stream()
new Stream()
new Stream()
new Stream()
new Stream()
new Stream()
new Stream()

% qscalac -optimize -Yinline-warnings -Xfatal-warnings sandbox/test.scala && echo ':javap -c Test$#test' | qscala && qscala -nc Test
Welcome to Scala version 2.12.0-20150204-161502-f894c55ef3 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_25).
Type in expressions to have them evaluated.
Type :help for more information.

scala> :javap -c Test$#test
  public void test();
    Code:
       0: new           #22                 // class Test$Stream
       3: dup
       4: invokespecial #23                 // Method Test$Stream."<init>":()V
       7: astore_1
       8: new           #25                 // class Test$StreamWithFilter
      11: dup
      12: aload_1
      13: invokespecial #28                 // Method Test$StreamWithFilter."<init>":(LTest$Stream;)V
      16: invokevirtual #32                 // Method Test$StreamWithFilter.filtered:()LTest$Stream;
      19: invokevirtual #35                 // Method Test$Stream.next:()LTest$Stream;
      22: invokevirtual #35                 // Method Test$Stream.next:()LTest$Stream;
      25: invokevirtual #35                 // Method Test$Stream.next:()LTest$Stream;
      28: invokevirtual #35                 // Method Test$Stream.next:()LTest$Stream;
      31: invokevirtual #35                 // Method Test$Stream.next:()LTest$Stream;
      34: invokevirtual #35                 // Method Test$Stream.next:()LTest$Stream;
      37: invokevirtual #35                 // Method Test$Stream.next:()LTest$Stream;
      40: invokevirtual #35                 // Method Test$Stream.next:()LTest$Stream;
      43: pop
      44: return
}

scala> :quit
new Stream()
new Stream()
new Stream()
java.lang.OutOfMemoryError: Java heap space
    at scala.reflect.ManifestFactory$$anon$6.newArray(Manifest.scala:93)
    at scala.reflect.ManifestFactory$$anon$6.newArray(Manifest.scala:91)
```

The inliner introduces a local variable where there was none before, which retains the head of the stream. I have lodged [SI-9137](https://issues.scala-lang.org/browse/SI-9137) to track this more general problem.

A way forward for this pull request might be to mark `Stream#withFilter` with the `@noinline` annotation to prevent it from being inlined into the test (or any other call site.)

retronym commented:
/cc @lrytz Have you seen a ticket related to this already? Or perhas a reference to this issue in some of the inlining literature you've been ploughing through?

Ichoran commented:
Nice detective work, @retronym!  I agree that `@noinline` should do the trick.  I wonder if this issue may affect other lazy computations and we should try to find and `@noinline` them all?  (Better yet would be a directive to `@nullout` which would allow inlining but would null out local variables as soon as possible.  I wonder how expensive that would be to do by default?)

ms-tg commented:
Thank you @retronym! That is awesome. I will make all code changes suggested so far today and re-push the PR, including the `@noinline` annotation on `#withFilter`.

ms-tg commented:
@retronym Progress! the `@noinline` trick definitely worked for test `withFilter_after_first_foreach_allows_GC`. However, it doesn't seem to have been enough for the test of `#withFilter.withFilter`, named `withFilter_after_first_withFilter_foreach_allows_GC`: https://scala-ci.typesafe.com/job/scala-2.11.x-validate-test/ws/build/junit/TEST-scala.collection.immutable.StreamTest.txt

```
Testsuite: scala.collection.immutable.StreamTest
Tests run: 6, Failures: 1, Errors: 0, Time elapsed: 1.883 sec

Testcase: foreach_allows_GC took 0.076 sec
Testcase: t6727_and_t6440 took 0.064 sec
Testcase: withFilter_can_retry_after_exception_thrown_in_filter took 0.003 sec
Testcase: withFilter_after_first_foreach_allows_GC took 0.018 sec
Testcase: withFilter_after_first_withFilter_foreach_allows_GC took 1.672 sec
    FAILED
null
junit.framework.AssertionFailedError
    at scala.collection.immutable.StreamTest.assertStreamOpAllowsGC(StreamTest.scala:42)
    at scala.collection.immutable.StreamTest.withFilter_after_first_withFilter_foreach_allows_GC(StreamTest.scala:62)

Testcase: filter_all_foreach_allows_GC took 0.036 sec
```

Any idea what's going on in that case? @Ichoran?

ms-tg commented:
Finally got all tests passing! Hurray! So I rebased to a single commit again, for your reviewing pleasure. 
@Ichoran @adriaanm @retronym @paulp

adriaanm commented:
Huzzah!

ms-tg commented:
@retronym ping

retronym commented:
Sorry to play ping/pong here, but I'm going to defer the remaining questions about the sufficient laziness to @Ichoran. I understand he has a little more time now.

ms-tg commented:
Ok thanks @retronym will follow up with him.

Hi @Ichoran, please let me know if you'd like to schedule an hour to discuss and review over voice or hangout? Can we review the old design, what was wrong (GC was not possible using `withFilter`), the design of this proposed solution, etc?

Ichoran commented:
I'll look at this later today.  I prefer to _not_ start out with a live review, now that things are working, because (1) live chat is a good way for people to share misconceptions, and (2) there is more of a time pressure associated with it which may not encourage adequate care in places that might need it.  After I've had a look if there are additional things to discuss (that I can't figure out from looking at the code plus the comments I've read on the PRs) we can figure out a way to have that discussion (e.g. by scheduling a time to have it live).

ms-tg commented:
Sounds good.

Ichoran commented:
@ms-tg - Sorry, didn't get a chance last night.  I'll do it tonight.

Ichoran commented on test/junit/scala/collection/immutable/StreamTest.scala:
```diff
@@ -15,4 +18,58 @@ class StreamTest {
+    val msgSuccessGC = "GC success"
+    val msgFailureGC = "GC failure"
+
+    val ref = WeakReference( Stream.from(1).take(500) ) // wait about 5 seconds to fail
```
Why is the comment "wait about 5 seconds to fail" here?

ms-tg replied:
The `.take(500)`, combined with the `Thread.sleep(10)` below, limits failures of these tests to a max of 5 seconds each. The alternative of course is to have the stream grow forever, never collecting, until the test framework times out or memory is exhausted.

Ichoran replied:
Maybe explaining that you will wait up to 5 seconds because you GC and then wait 10 ms up to 500x is a good idea.  The logic is so spread out between the WeakReference and the `gcAndThrowIfCollected` method and the other logic that it's not obvious that this is what happens.

Ichoran commented on test/junit/scala/collection/immutable/StreamTest.scala:
```diff
@@ -15,4 +18,58 @@ class StreamTest {
+
+    def gcAndThrowIfCollected(n: Int): B = {
+      System.gc()
+      Thread.sleep(10)
```
Why sleep for 10 ms?

ms-tg replied:
(please see above response regarding the `.take(500)`)

Ichoran commented on test/junit/scala/collection/immutable/StreamTest.scala:
```diff
@@ -15,4 +18,58 @@ class StreamTest {
+  /** Test helper to verify that the given Stream operation allows
+    * GC of the head during processing of the tail.
+    */
+  def assertStreamOpAllowsGC[B](op: (=> Stream[Int], Int => B) => Any, f: Int => B): Unit = {
```
What is the generic doing here?  You always seem to use it as a no-op.

ms-tg replied:
It is making it possible, in a later commit, for me to retrofit test coverage of `.map` and `.flatMap` also allowing GC, in both the regular Stream and `.withFilter` varieties.

(To be clear, this is not implemented yet. My plan was to get an initial fix of this issue accepted, and then to do follow on PR filling in the regression tests of GC, so that ultimately the fact that Stream's methods allow GC becomes a feature that is covered by tests on an on-going basis)

Ichoran replied:
You can always change this later when you do the retrofit.  Why not keep it as simple as possible now, so it's easier to verify correctness, and then retrofit this when you retrofit the others?

Ichoran commented on test/junit/scala/collection/immutable/StreamTest.scala:
```diff
@@ -15,4 +18,58 @@ class StreamTest {
+  @Test // SI-8990
+  def withFilter_can_retry_after_exception_thrown_in_filter: Unit = {
+    var shouldThrow = true
+    def greaterThanFiveButMayThrow(n: Int) = if (shouldThrow && n == 5) throw new RuntimeException("n == 5") else n > 5
```
I'm not sure the really long method name here really clarifies things.  Why not just put the whole body into the filter and use a comment to explain it if you need to?

ms-tg replied:
I'm not certain what you mean here. I have no doubt that you are more intelligent than I am, but even as the author of these tests, I find the names very useful in remembering what is going on in them and what the key points are?

Ichoran replied:
The problem is that `greaterThanFiveButMayThrow` does something unusual, and the name despite being very long doesn't adequately explain either what it does (i.e. that it throws on 5) or why you want it.  If you put the body directly into the `withFilter` statement, you can explain in a comment that you are checking that an intermittent exception will cause re-evaluation of the filter statement.  And you don't really need a named val because you only use it once.

Ichoran commented on test/junit/scala/collection/immutable/StreamTest.scala:
```diff
@@ -15,4 +18,58 @@ class StreamTest {
+      System.gc()
+      Thread.sleep(10)
+      if (ref.get.isEmpty) throw new RuntimeException(msgSuccessGC)
+      assertTrue(n >= 1) // protect from JIT optimizations
```
What happens in optimization, and how does this help?  Do you need both this and `f(n)`?

ms-tg replied:
Perhaps we can remove that line now. It was added in response to an earlier suggestion, I've forgotten from whom. However, the inlining that ultimately was the issue was at the call to `.withFilter`, and the `@noinline` was the fix, so this may be doing nothing?

Ichoran replied:
Why not try removing it and see if it still works.  It's really helpful to have all the code either have an obvious purpose, or a comment explaining why this particular thing is necessary.  "Someone suggested this" is a little dodgy either way.

Ichoran commented on test/junit/scala/collection/immutable/StreamTest.scala:
```diff
@@ -15,4 +18,58 @@ class StreamTest {
+    var shouldThrow = true
+    def greaterThanFiveButMayThrow(n: Int) = if (shouldThrow && n == 5) throw new RuntimeException("n == 5") else n > 5
+    val wf = Stream.from(1).take(10).withFilter(greaterThanFiveButMayThrow)
+    Try { wf.map(identity) }                   // throws on n == 5
```
Shouldn't you check that it actually does throw?

ms-tg replied:
Can you explain a bit of your reasoning and intent in this question? I'm not sure what you are implying -- that the test would check that n == 5 was ever detected?

Ichoran replied:
Right.  You _claim_ that it will throw, and the test only tests something interesting if it _does_ throw, but in a unit test you should probably check that your expectations are met (i.e. that it did throw something).

Ichoran commented:
The change to withFilter looks good but the test will need another pass or two (one to inform me of what the intent is, and a second to either put that into comments or change how things are implemented)

ms-tg commented:
Thanks for your review. I couldn't respond for a few days as we were hurriedly moving apartments, a challenging task even when one's wife isn't 8 months pregnant ;). However, I am now at least somewhat moved in and ready to finish this off with you.

ms-tg commented:
Hi @Ichoran, I've pushed a commit which I think addresses all the concerns you raised above. When you have a moment, it would be great if you could review again.

ms-tg commented:
@Ichoran ping -- can you review whether all raised concerns about the tests are addressed? Avail for live discussion if/when you'd like.

Ichoran commented:
LGTM - thanks @ms-tg for tackling this deceptively simple-seeming issue and taking it through to a good solution!

ms-tg commented:
@Ichoran -- very cool. Do you need me to document anywhere how the `Stream#withFilter` now works? Also, do you need me to re-squash this PR so you can merge as one commit?

Ichoran commented:
@ms-tg - Yes, please squash to a single commit.  Since the new behavior is closer to the expected behavior, I don't think it needs to be documented beyond what you've already done.  It obeys the API, comes closer to the point of what Stream is for, and anyone who really cares about internals can find ample hints/comments in the code.

Ichoran commented:
LGTM

ms-tg commented:
Ok, squashed! Thanks for all your help. Can't wait to see my first commit merged into Scala ;)

ms-tg commented:
ping?

adriaanm commented:
Merging! :fireworks: Epic entrance, @ms-tg! :fireworks: 

May the next one be a smoother experience :-)

ms-tg commented:
:sunny: thanks for all the help!
